README.md
# dialog.js #

[![Codacy Badge](https://api.codacy.com/project/badge/Grade/a61ce490dd304193865a6b93f793e8b4)](https://www.codacy.com/app/mttbernardini/dialog?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=mttbernardini/dialog&amp;utm_campaign=Badge_Grade)

This project is a light JS library providing basic asynchronous modal dialogs for webpages, aiming to replace the built-in `alert()`, `prompt()` and `confirm()` which instead block JavaScript execution.

The library provides one function `dialog()` in the global namespace (`window`), taking as argument an object of properties and returning a `Promise` resolving to the result object.

```js
prop = {
	type: "alert",
	title: "Message",
	content: "Hello, world!",
};

dialog(prop).then(myFunc);
// Shows the dialog box and calls `myFunc` after closing it
```


## How to use ##

The inclusion on a project is straightforward and only consist of adding the script and the default stylesheet:

```html
<link rel="stylesheet" type="text/css" href="/path/to/dialog_default.css">
<script type="text/javascript" src="/path/to/dialog.js"></script>
```


## Documentation ##

A short and effective documentation for all the supported properties can be found at the beginning of the `dialog.js` file.

A more detailed documentation providing examples can be found in the [wiki][1]


## Live demo ##

You can see three live examples [here][2].


## Compatibility ##

The software is compatible with all major browsers. A polyfill might be needed in order to support `Promise`s, no other ES6 features are used.

The stylesheet is W3C valid. A prefixer (like [-prefix-free][3]) might be needed in order to add vendor prefixes.


## License and sharing ##

© 2015 Matteo Bernardini, [@mttbernardini][4].

This project is licensed under the MIT License.  
Please refer to the LICENSE file for further information.


[1]: https://github.com/mttbernardini/dialog/wiki
[2]: https://mttbernardini.github.io/dialog/demo.html
[3]: https://github.com/LeaVerou/prefixfree
[4]: https://twitter.com/mttbernardini


changelog.md
# Changelog of *dialog.js* #

## Version 3 ##

### 3.0 ###

*Released on 2017-09-10*

- Removed polyfill for `addEventListener` on IE8.
- `dialog()` now returns a `Promise` (no more callbacks) resolving to the result object (previously known as `returnObj`).
- Renamed `vars` → `data`, both in the options for `dialog` and in the `Promise` resolved object.

NB: These changes break compatibility with v2.x. Please read the new documentation for further details.

NB: This version drops support for older browsers and IE. Consider using polyfills for `addEventListener` and `Promise` for browsers not supporting them.


## Version 2 ##

### 2.1 ###

*Released on 2017-07-14*

- `returnObj.value` is correctly set to `undefined` if a dialog is not of "prompt" type.
- Renamed button classes (`dialog-continue-button` → `dialog-continue-btn`, `dialog-cancel-button` → `dialog-cancel-btn`).
- Removed unneeded `data-action` attributes on dialog buttons.
- Simplified code (increased readability).

### 2.0.1 ###

*Released on 2017-02-18*

- Minor bugfixes.

### 2.0.0 ###

*Released on 2015-10-18*

- Reconception of the majority of the code.
- Uploaded the project on GitHub.
- Removed `data-*` attributes for styling, now using classes (refer to style.css file to see new classes used).
- Renamed callBack → callback, which is now a reference to a function (with `returnObj` passed as first argument), not an executable string (not `eval` anymore!).
- Renamed returnObj.boolean → returnObj.action (same behavior).
- Title and content can't contain HTML anymore. However LF characters and whitespaces are preserved.
- Adding to the DOM only the visible elements (removed `display: none` hack).
- Fixed a bug with the `keyup` event listener.

NB: These changes break compatibility with v1.x. Please read the new documentation for further details.


## Version 1 ##

### 1.1 ###

*Released on 2012-13-09*

- Fixed a problem with the XHTML syntax. Now the function works good both in HTML and XHTML, automatically.

### 1.0 ###

*Released on 2012-08-07*

- This is the first version of the code, which includes the 3 objects `dialogSettings`, `dialog()` and `returnObj`.
- The code integrates the default stylesheet. The semitransparent background of the dialog doesn't work with IE < 9.
- When pressing <kbd>enter</kbd> the script will emulate the *ok* or *continue* button click, and the <kbd>esc</kbd> key will emulate the *cancel* button click when the cancel button is shown.


dialog.js
/*
 * dialog.js is a utility which can replace the default JavaScript dialog windows for webpages (alert(), prompt(), confirm())
 *
 * Version: 3.0
 * Copyright (c) 2015 Matteo Bernardini
 * Licensed under the MIT License (refer to the LICENSE file for further information).
 *
 * Check this project on GitHub:
 * https://github.com/mttbernardin/dialog
 *
 */

/**
 * The library provides 3 types of objects.
 *
 * ## A global Object `dialogSettings`, used to modify the default behavior of the dialog boxes ##
 *
 * __object_key__   __type__   __default_value__   __comment__
 * defType          string     "alert"             Default type
 * defTitle         string     "Message"           Default title
 * defContent       string     "Missing text"      Default message of the dialog
 * okText           string     "OK"                Label for the <ok> button
 * continueText     string     "Continue"          Label for the <continue> button
 * cancelText       string     "Cancel"            Label for the <cancel> button
 *
 *
 * ## The main Function `dialog(opt)`, returns a Promise and takes the following options as properties of the Object `opt` ##
 *
 * __object_key__   __type__   __comment__
 * type             string     Type of the dialog window. Possible values are ['alert', 'prompt', 'confirm']
 * title            string     Text which will be displayed as the title of the dialog window
 * content          string     Message of the dialog window
 * placeholder      string     If the type is 'prompt', this will define a placeholder text for the input box
 * id               string     Identifier which can be referenced afterwards in the resolved object (see below)
 * data             Object     Object of convenience that will be referenced in the resolved object (see below)
 *
 *
 * ## The resolved Object of the Promise ##
 *
 * __object_key__   __type__   __comment__
 * id               string     Identifier passed when calling the `dialog()` function, defaults to `undefined`
 * data             Object     Reference of the object passed with the `dialog()` function, defaults to `undefined`
 * action           boolean    `true` if the user has pressed <ok> or <continue>, `false` for <cancel>
 * value            string     If the type of the requested dialog window was 'prompt' it will contain
 *                             the value inserted in the input box, otherwise it will be `undefined`
 *
 */


"use strict";

var dialogSettings = new Object();

function dialog(params) {

	params         = params         || new Object();
	dialogSettings = dialogSettings || new Object();

	// == Default Values == //
	dialogSettings.defTitle     =  dialogSettings.defTitle      ||  "Message";
	dialogSettings.defType      =  dialogSettings.defType       ||  "alert";
	dialogSettings.defContent   =  dialogSettings.defContent    ||  "Missing text";
	dialogSettings.okText       =  dialogSettings.okText        ||  "OK";
	dialogSettings.continueText =  dialogSettings.continueText  ||  "Continue";
	dialogSettings.cancelText   =  dialogSettings.cancelText    ||  "Cancel";

	// == Switch between default values or passed values == //
	params.title    =  params.title    ||  dialogSettings.defTitle;
	params.type     =  params.type     ||  dialogSettings.defType;
	params.content  =  params.content  ||  dialogSettings.defContent;


	// == Creating elements == //
	var parts = ["window", "wrapper", "title", "body", "message", "prompt", "actions", "ok-btn", "cancel-btn"],
	    elms  = {};

	for (var i = 0, partName; partName = parts[i]; i++) {
		if (partName.indexOf("btn") !== -1) {
			elms[partName] = document.createElement("button");
			elms[partName].setAttribute("type", "button");
		} else {
			elms[partName] = document.createElement("div");
		}
		elms[partName].className = "dialog-" + partName;
	}

	var prompt = document.createElement("input");
	prompt.setAttribute("type", "text");
	prompt.className = "dialog-prompt-input";


	// == Arranging elements == //
	elms["window"].appendChild(elms["wrapper"]);
	elms["wrapper"].appendChild(elms["title"]);
	elms["wrapper"].appendChild(elms["body"]);
	elms["body"].appendChild(elms["message"]);
	elms["actions"].appendChild(elms["ok-btn"]);

	switch (params.type) {
		case "prompt":
			elms["prompt"].appendChild(prompt);
			elms["body"].appendChild(elms["prompt"]);
		case "confirm": // or "prompt", notice no break here
			elms["actions"].appendChild(elms["cancel-btn"]);
	}

	elms["body"].appendChild(elms["actions"]);


	// == Assigning values == //
	elms["title"].textContent       =  params.title;
	elms["message"].textContent     =  params.content;
	elms["ok-btn"].textContent      =  params.type === "confirm" ? dialogSettings.continueText : dialogSettings.okText;
	elms["cancel-btn"].textContent  =  dialogSettings.cancelText;
	prompt.placeholder              =  params.placeholder || "";


	// == Appending to body & focus == //
	var dialogWindow = elms["window"];
	dialogWindow.setAttribute("data-dialog-type", params.type);

	document.body.appendChild(dialogWindow);

	if (params.type === "prompt")
		prompt.focus();



	// == Return promise == //
	return new Promise(function(resolve, reject) {

		// Will be triggered when closing the dialog
		function action(e, userAction) {
			e = e || window.event;
			var target = e.target ? e.target : e.srcElement;

			if (typeof userAction === "boolean" || target.tagName.toLowerCase() === "button") {
				document.removeEventListener("keyup", useKeys, false);
				document.body.removeChild(dialogWindow);
				resolve({
					id:     params.id,
					action: typeof userAction === "boolean" ? userAction : target.className.indexOf("ok") !== -1,
					value:  params.type === "prompt" ? prompt.value : void 0,
					data:   params.data
				});
			}
		}

		// Handler for keyboard events
		function useKeys(e) {
			e = e || window.event;
			if (e.keyCode === 13) // ENTER key
				action(null, true);
			else if (e.keyCode === 27 && params.type !== "alert") // ESC key
				action(null, false);
		}

		document.addEventListener("keyup", useKeys, false);
		elms["actions"].addEventListener("click", action, false);

	});
}


script.js
/*
 *
 * Asynchronous Modal Dialog Window, a utility which can replace the default JavaScript dialog windows for webpages (alert(), prompt(), confirm())
 *
 * The program integrates 3 elements.
 *
 * ## The first is a global Object named `dialogSettings` that you can use to modify the default behavior of the dialog windows ##
 *
 * __object_key__   __type__   __default_value__       __comment__
 * defType          string     "alert"                 default type
 * defTitle         string     "Message"               default title
 * defContent       string     "Missing text"          default message of the dialog
 * okText           string     "OK"                    default content of the <ok> button
 * continueText     string     "Continue"              default content of the <continue> button
 * cancelText       string     "Cancel"                default content of the <cancel> button
 *
 *
 * ## The second element is the main Function, `dialog()`, which needs the following options, passed with an object as first argument ##
 *
 * __object_key__   __type__   __comment__
 * type             string     type of the dialog window. Possible values are ['alert', 'prompt', 'confirm']
 * title            string     text which will be displayed as the title of the dialog window
 * content          string     message of the dialog window
 * placeholder      string     if the type is 'prompt', this will define a placeholder text for the input box
 * id               string     identifier which can be retrieved afterwards in the `returnObj`, on the callback function
 * vars             Object     object of convenience that will be returned in the `returnObj`, on the callback function
 * callback         Function   function to be called when the dialog window is closed
 *
 *
 * ## The last one is an Object that will be returned in the callback function as first argument, named `returnObj` for convenience of this guide ##
 *
 * __object_key__   __type__   __comment__
 * id               string     identifier passed when calling the `dialog()` function, defaults to undefined
 * vars             Object     same object passed with the function (NB: same reference), defaults to undefined
 * action           boolean    `true` if the user has pressed <ok> or <continue>, `false` for <cancel>
 * value            string     if the type of the requested dialog window was "prompt" it will contain the value inserted in the input box, otherwise it will be undefined
 *
 * ================================================================================================
 *
 * Version: 2.0.1
 * Copyright (c) 2015 Matteo Bernardini
 * Licensed under the MIT License (refer to the LICENSE file for further information).
 *
 * Check this project on GitHub:
 * http://github.com/mttbernardin/dialog
 *
 */



// Backwards compatibility with IE < 9
if (!window.addEventListener) {
	Object.defineProperties(Element.prototype, {
		"addEventListener": {
			value: function(evt, handler, capture) {
				this.attachEvent("on"+evt, handler);
			}
		},
		"removeEventListener": {
			value: function(evt, handler, capture) {
				this.detachEvent("on"+evt, handler);
			}
		},
	});
}


// MAIN FUNCTION //

var dialogSettings = new Object();

function dialog(params) {

	params         = params         || new Object();
	dialogSettings = dialogSettings || new Object();

	// == Default Values == //
	dialogSettings.defTitle     =  dialogSettings.defTitle      ||  "Message";
	dialogSettings.defType      =  dialogSettings.defType       ||  "alert";
	dialogSettings.defContent   =  dialogSettings.defContent    ||  "Missing text";
	dialogSettings.okText       =  dialogSettings.okText        ||  "OK";
	dialogSettings.continueText =  dialogSettings.continueText  ||  "Continue";
	dialogSettings.cancelText   =  dialogSettings.cancelText    ||  "Cancel";

	// == Switch between default values or passed values == //
	params.title    =  params.title    ||  dialogSettings.defTitle;
	params.type     =  params.type     ||  dialogSettings.defType;
	params.content  =  params.content  ||  dialogSettings.defContent;


	// == Creating elements == //
	var parts = ["window", "wrapper", "title", "body", "message", "prompt", "actions", "ok-btn", "cancel-btn"],
	    elms  = {};

	for (var i = 0, partName; partName = parts[i]; i++) {
		if (partName.indexOf("btn") !== -1) {
			elms[partName] = document.createElement("button");
			elms[partName].setAttribute("type", "button");
		} else {
			elms[partName] = document.createElement("div");
		}
		elms[partName].className = "dialog-" + partName;
	}

	var prompt = document.createElement("input");
	prompt.setAttribute("type", "text");
	prompt.className = "dialog-prompt-input";


	// == Arranging elements == //
	elms["window"].appendChild(elms["wrapper"]);
	elms["wrapper"].appendChild(elms["title"]);
	elms["wrapper"].appendChild(elms["body"]);
	elms["body"].appendChild(elms["message"]);
	elms["actions"].appendChild(elms["ok-btn"]);

	switch (params.type) {
		case "prompt":
			elms["prompt"].appendChild(prompt);
			elms["body"].appendChild(elms["prompt"]);
		case "confirm": // or "prompt", notice no break here
			elms["actions"].appendChild(elms["cancel-btn"]);
	}

	elms["body"].appendChild(elms["actions"]);


	// == Assigning values == //
	elms["title"].textContent       =  params.title;
	elms["message"].textContent     =  params.content;
	elms["ok-btn"].textContent      =  params.type === "confirm" ? dialogSettings.continueText : dialogSettings.okText;
	elms["cancel-btn"].textContent  =  dialogSettings.cancelText;
	prompt.placeholder              =  params.placeholder || "";


	// == Appending to body & focus == //
	var dialogWindow = elms["window"];
	dialogWindow.setAttribute("data-dialog-type", params.type);

	document.body.appendChild(dialogWindow);

	if (params.type === "prompt")
		prompt.focus();



	// == Adding event listeners == //

	// Will be triggered when closing the dialog
	function action(e, userAction) {
		e = e || window.event;
		var target = e.target ? e.target : e.srcElement;

		if (typeof userAction === "boolean" || target.tagName.toLowerCase() === "button") {
			document.removeEventListener("keyup", useKeys, false);
			document.body.removeChild(dialogWindow);
			//Callback function if defined
			if (params.callback) {
				var returnObj = {
					id: params.id,
					action: typeof userAction === "boolean" ? userAction : target.className.indexOf("ok") !== -1,
					value: params.type === "prompt" ? prompt.value : void 0,
					vars: params.vars
				};
				params.callback(returnObj);
			}
		}
	}

	// Handler for keyboard events
	function useKeys(e) {
		e = e || window.event;
		if (e.keyCode === 13) // ENTER key
			action(null, true);
		else if (e.keyCode === 27 && params.type !== "alert") // ESC key
			action(null, false);
	}

	document.addEventListener("keyup", useKeys, false);
	elms["actions"].addEventListener("click", action, false);

}
